### PR TITLE
test: Run tests faster with `nextest`

### PR DESCRIPTION
**Pull request by Hywan**
> [`cargo-nextest`](https://nexte.st/index.html) is a next-generation
> test runner for Rust projects.

This patch installs and uses `nextest` to run our own tests.

Comparing `cargo test` and `cargo nextest` with hyperfine provides the
following results:

```
$ hyperfine 'cargo test --workspace' 'cargo nextest run --workspace && cargo test --doc'
Benchmark 1: cargo test --workspace
  Time (mean ± σ):     51.785 s ±  2.066 s    [User: 183.471 s, System: 10.563 s]
  Range (min … max):   49.151 s … 56.641 s    10 runs

Benchmark 2: cargo nextest run --workspace && cargo test --doc
  Time (mean ± σ):     44.556 s ±  0.894 s    [User: 192.213 s, System: 11.441 s]
  Range (min … max):   43.170 s … 45.762 s    10 runs
```

Benchmark 2 is 1.16 times faster than Benchmark 1

### CODE REVIEW

**Review by jplatte — COMMENTED** [review]
Are the figures from the first commit description up-to-date now?

File: xtask/src/ci.rs [hunk 1]
```diff
@@ -141,12 +137,6 @@ fn check_docs() -> Result<()> {
     build_docs([], DenyWarnings::Yes)
 }
 
-fn run_tests() -> Result<()> {
-    cmd!("rustup run stable cargo test").run()?;
-    cmd!("rustup run beta cargo test").run()?;
-    Ok(())
-}
```
[the remark this answers is not in the document]
Comment (Hywan) [reply xtask/src/ci.rs:148]: It is indeed unused. I'll update.

**Comment by Hywan** [comment]
With `xargs -P`, we can achieve the following results:

```
Benchmark 1: cargo test --workspace
  Time (mean ± σ):     51.785 s ±  2.066 s    [User: 183.471 s, System: 10.563 s]
  Range (min … max):   49.151 s … 56.641 s    10 runs

Benchmark 2: cargo nextest run --workspace && cargo test --doc
  Time (mean ± σ):     44.556 s ±  0.894 s    [User: 192.213 s, System: 11.441 s]
  Range (min … max):   43.170 s … 45.762 s    10 runs

Benchmark 3: echo "cargo nextest run --workspace\0cargo test --doc" | xargs -0 -P 0 -I@ sh -c @
  Time (mean ± σ):     34.339 s ±  4.357 s    [User: 192.472 s, System: 11.739 s]
  Range (min … max):   22.009 s … 36.422 s    10 runs
```

Benchmark 2 is 1.16 times faster than Benchmark 1
Benchmark 3 is 1.51 times faster than Benchmark 1

I'm not sure how it is possible to run such `xargs` command in a YAML file while having cross-compatibility, but let's try!

**Comment by Hywan** [comment]
Without `xargs -P`, here are the results for the “test” step of the CI workflow for some jobs (I won't do it exhaustively, time differs, it's just to provide an order of magnitude idea):

| jobs | before | after | speed up factor |
|--|--|--|--|
| linux / stable | 11m 21s | 9m 13s | **1.23** |
| macOS / stable | 15m 25s | 10m 40s | **1.45** |
| linux / features-sso-login | 9m 30s | 5m 9s | **1.84** |
| linux / features-rustls-tls | 5m 4s | 2m 13s | **2.29** |
| linux / features-no-encryption | 6m 16s | 2m 14s | **2.81** |
| linux / features-no-sled | 6m 51s | 4m 9s | **1.65** |
| linux / features-no-encryption-and-sled | 5m 10s | 1m 51s | **2.79** |

Just on those examples, running tests is 2.01 times faster on average.

**Comment by Hywan** [comment]
When running `cargo nextest run` and `cargo test --doc` in parallel, we have worst results on Github Actions, probably because the number of threads is limited. In that case, they are spending more time waiting on each other than truly running what we want. Let's keep things simple and let's use `cargo-nextest` for now.

**Comment by Hywan** [comment]
> Removal of dead code is still not a separate commit afaict but otherwise looks good.

Done, https://github.com/matrix-org/matrix-rust-sdk/pull/778/commits/d9475c131a43c1b08657804a1ec365492e7b6926 :-).

**Pull request merged**